tf_quant_finance/math/jacobian_test.py

```diff
@@ -0,0 +1,120 @@
+# Lint as: python3
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for math.jacobian.py."""
+from functools import partial
+
+from absl.testing import parameterized
+
+import numpy as np
+import tensorflow.compat.v2 as tf
+from tensorflow.python.framework import test_util  # pylint: disable=g-direct-tensorflow-import
+
+import tf_quant_finance as tff
+
+
+def quadratic(p, x):
+  """Quadratic function."""
+  a = tf.expand_dims(p[..., 0], axis=-1)
+  b = tf.expand_dims(p[..., 1], axis=-1)
+  c = tf.expand_dims(p[..., 2], axis=-1)
+  return a * x**2 + b * x + c
+
+
+@test_util.run_all_in_graph_and_eager_modes
+class JacobianTest(parameterized.TestCase, tf.test.TestCase):
+  """Jacobian test cases."""
+
+  @parameterized.named_parameters(
+    {
+      'testcase_name': 'SinglePrecision',
+      'dtype': tf.float32
+    }, {
+      'testcase_name': 'DoublePrecision',
+      'dtype': tf.float64
+    },
+  )
+  def test_jacobian(self, dtype):
+    """Test function jacobian."""
+    # Shape [2]
+    x = tf.range(1, 3, dtype=dtype)
+    func = partial(quadratic, x=x)
+    with self.subTest("SingleTensor"):
+      # Shape [3]
+      ps = tf.constant([1.0, 2.0, -1.0], dtype=dtype)
+      # Shape [2, 3]
+      expected_jacobian = [[1.0, 1.0, 1.0], [4.0, 2.0, 1.0]]
+      jacobian = self.evaluate(tff.math.jacobian(func, ps))
+      self.assertEqual(jacobian.shape, (2, 3))
+      np.testing.assert_allclose(jacobian, expected_jacobian)
+    with self.subTest("BatchedTensor"):
+      # Shape [2, 3]
+      ps = tf.constant([[1.0, 2.0, -1.0], [2.0, 0.0, 0.0]], dtype=dtype)
+      # Shape [2, 3]
+      expected_jacobian = [
+        [[1.0, 1.0, 1.0], [4.0, 2.0, 1.0]],
+        [[1.0, 1.0, 1.0], [4.0, 2.0, 1.0]],
+      ]
+      jacobian = self.evaluate(tff.math.jacobian(func, ps))
+      self.assertEqual(jacobian.shape, (2, 2, 3))
+      np.testing.assert_allclose(jacobian, expected_jacobian)
+
+  @parameterized.named_parameters(
+    {
+      'testcase_name': 'SinglePrecision',
+      'dtype': tf.float32
+    }, {
+      'testcase_name': 'DoublePrecision',
+      'dtype': tf.float64
+    },
+  )
+  def test_value_and_jacobian(self, dtype):
+    """Test function value_and_jacobian."""
+    # Shape [2]
```

Comment:
rank?

Reply:
Hm...I guess shape = 2 is correct? with rank = 1?